ksonnet-gen/printer/printer.go

```diff
@@ -0,0 +1,345 @@
+		}
+	case *ast.LiteralString:
+		var quoteType byte
+
+		switch t.Kind {
+		default:
+			return nil, errors.Errorf("unknown string literal kind %#v", t.Kind)
+		case ast.StringDouble:
+			quoteType = doubleQuote
+		case ast.StringSingle:
+			quoteType = singleQuote
+		}
+
+		buf.Write([]byte{quoteType})
+		buf.WriteString(t.Value)
+		buf.Write([]byte{quoteType})
+	case *ast.LiteralNumber:
+		buf.WriteString(t.OriginalString)
+	case *ast.Self:
+		buf.WriteString("self")
+	case *ast.Var:
+		buf.WriteString(string(t.Id))
+	}
+
+	return buf.Bytes(), nil
+}
+
+func bopToString(bop ast.BinaryOp) (string, error) {
+	for k, v := range ast.BopMap {
+		if v == bop {
+			return k, nil
+		}
+	}
+
+	return "", errors.New("unknown binary operation")
+}
+
+func (p *printer) handleApply(t *ast.Apply, buf *bytes.Buffer) error {
+	s, err := extractApply(t.Target)
+	if err != nil {
+		return err
+	}
+	buf.WriteString(s)
+	buf.WriteString("(")
+	b, err := p.output(t.Arguments)
+	if err != nil {
+		return err
+	}
+	buf.Write(b)
+	buf.WriteString(")")
+
+	return nil
+}
+
+func (p *printer) handleObjectField(t ast.ObjectField, buf *bytes.Buffer) error {
+	if t.Comment != nil {
+		isNew := true
+
+		buf.WriteString("//")
+		for _, r := range t.Comment.Text {
+			if r == rune(newline) {
+				buf.Write([]byte{newline})
+				buf.WriteString("//")
+				isNew = true
+				continue
+			}
+			if isNew {
+				buf.Write([]byte{blank})
+				isNew = false
+			}
+
+			buf.Write([]byte{byte(r)})
+		}
+		buf.Write([]byte{newline})
+	}
+
+	var fieldType string
+
+	switch hide := t.Hide; hide {
+	default:
+		return errors.Errorf("unknown Hide type %#v", hide)
+	case ast.ObjectFieldHidden:
+		fieldType = "::"
+	case ast.ObjectFieldVisible:
+	case ast.ObjectFieldInherit:
+		fieldType = ":"
+	}
+
+	switch kind := t.Kind; kind {
+	default:
+		return errors.Errorf("unknown Kind type %#v", kind)
+	case ast.ObjectFieldID:
+		buf.WriteString(string(*t.Id))
+		if t.Method != nil {
+			addMethod(t, buf)
+		}
+
+		if t.SuperSugar {
+			buf.Write([]byte{syntaxSugar})
+		}
+
+		buf.WriteString(fieldType)
+		buf.Write([]byte{blank})
+		b, err := p.output(t.Expr2)
+		if err != nil {
+			return err
+		}
+		buf.Write(b)
+	case ast.ObjectLocal:
+		buf.WriteString("local ")
+		buf.WriteString(string(*t.Id))
+		if t.Method != nil {
+			addMethod(t, buf)
+		}
+		buf.WriteString(" = ")
+		b, err := p.output(t.Expr2)
+		if err != nil {
+			return err
+		}
+		buf.Write(b)
+	}
+
+	return nil
+}
+
+func addMethod(of ast.ObjectField, buf *bytes.Buffer) {
+	buf.WriteString("(")
+	var args []string
+	for _, arg := range of.Method.Parameters.Required {
+		args = append(args, string(arg))
+	}
+	buf.WriteString(strings.Join(args, ", "))
+	buf.WriteString(")")
+
+}
+
+func extractApply(n ast.Node) (string, error) {
+	switch t := n.(type) {
+	default:
+		return "", errors.Errorf("invalid type %T when extracting apply", t)
+	case *ast.Index:
+		var s string
+		if t.Target != nil {
```

Comment:
Are we planning to support `t.Index` as well? It shoudl be basically the same as this code, so it seems like we can get it "for free".

Reply:
Index now supports Index and Id fields.